### DISCUSSION TITLE
Unique hash result for same ID values in different modules

### DISCUSSION THREAD

**Discussion by micobarac** [deligoez/laravel-model-hashid#13] [Ideas]
It would be nice to include model name or prefix in hashing logics, which would provide unique hash result for same ID values in different modules:

**Current comparison:**

**User:**

`id = 1`
`hashId = use_xLlBdZW7ZEqe7`

**Company**

`id = 1`
`hashid = com_xLlBdZW7ZEqe7`

As you can see, hash suffix is the same for both models.

**Suggestion for improvement:**

`userHashId = hash('use'.1);`

`companyHashId = hash('com'.1);
`

This would provide better hashing results.

**Comment by deligoez** [comment]
Hi @micobarac,

I did not quite understand what you mean, but you can already customize hash id generation per model.

Just look this section on the config file:
https://github.com/deligoez/laravel-model-hashid/blob/a8ea6575311ea12144debba7620a14c4f71a776e/config/model-hashid.php#L122-L141

So if you want to have different hash for the same id for different models, just define an other salt for the model.

You can customize prefix generation or define your own prefix for a model on this config section too.

This is all documented on [README](https://github.com/deligoez/laravel-model-hashid/blob/main/README.md) file

**Comment by micobarac** [comment]
Hi @deligoez ,

I have seen that configuration part. I must admit, it is a bit cumbersome to configure all module classes this way, especially in huge systems.

Using the default configuration, generated raw hashid values for the same ID in different module classes should not be the same.
Look the example I noted above.

Maybe, 3 character model prefix should be concatenated with salt before hashing, to randomize raw values a bit.

```php
$hashid = hash($value, $salt . $prefix);
```

**Reply by deligoez** [reply]
Hi @micobarac again,

I forget to answer 😅.

Now I understand your concert and I think you are right.

But to generate a hash id we need an integer value. [Hash ids](https://github.com/vinkla/hashids) work that way by implementation.

I'll add this case to my roadmap and think about a `secure` solution to solve this.

Thanks for noticing.

**Comment by deligoez** [comment]
This is a valid concern. The recommended solution is to use **per-model salts** via the `model_generators` config:

```php
'model_generators' => [
    App\Models\User::class => [
        'salt' => 'user-specific-salt',
    ],
    App\Models\Company::class => [
        'salt' => 'company-specific-salt',
    ],
],
```

With different salts, the same integer ID will produce completely different raw hash values for each model. This has been available since v2 and is well-documented in the [Model-Specific Configuration](https://github.com/deligoez/laravel-model-hashid#model-specific-configuration) section of the README.

The prefix already differentiates the full hash ID (`use_xxx` vs `com_xxx`), but if you want different raw values too, per-model salts are the way to go.
